examples/asa/assets/5-contract-owned-asa-approval.py

```diff
@@ -0,0 +1,80 @@
+from pyteal import *
+
+def contract_owned_asa():
+    '''
+    This program is stateful part of contract owned ASA, This program keeps track of
+    number of ASA created by the contract, It also checks for
+    : While payment only creator can receive the payments
+    : While asset creation only creator can create assets
+    Owner can also be changed using this program
+    '''
+
+    on_creation = Seq([
+        App.globalPut(Bytes("Creator"), Txn.sender()),
+        App.globalPut(Bytes("ASACounter"), Int(0)),
+        Return(Int(1))
+    ])
+
+    # verify neither transaction
+    # contains a rekey
+    rekey_check = And(
+        Gtxn[0].rekey_to() == Global.zero_address(),
+        Gtxn[1].rekey_to() == Global.zero_address()
+    )
+
+    on_asa_create = Seq([
+        Assert(And(
+            App.globalGet(Bytes("Creator")) == Txn.sender(),
+            App.globalGet(Bytes("ASACounter")) == Int(0), # limit no of asa from here: currently set to max 1 ASA.
+            Global.group_size() == Int(3),
+            Gtxn[1].type_enum() == TxnType.AssetConfig,
+            Gtxn[2].type_enum() == TxnType.Payment,
+            rekey_check,
+            Gtxn[2].rekey_to() == Global.zero_address(),
+            Gtxn[2].amount() == Int(1000000)
+        )),
+        App.globalPut(Bytes("ASACounter"), Int(1)),
+        Return(Int(1))
+    ])
+
+    on_pay = Seq([
+        Assert(And(
+            App.globalGet(Bytes("Creator")) == Txn.sender(),
+            App.globalGet(Bytes("Creator")) == Gtxn[1].asset_receiver(),
+            rekey_check,
+        )),
+        Return(Int(1))
+    ])
+
+    # Change owner of the contract if called by current owner
+    change_owner = Seq([
+        Assert(And(
+            Txn.application_args[0] == Bytes("changeowner"),
+            App.globalGet(Bytes("Creator")) == Txn.sender()
+        )),
+        App.globalPut(Bytes("Creator"), Txn.application_args[1]),
```

Comment:
Also let's put in a note that with *TEALv3* we can get address from `appAccounts` as well (Using `Txn.Accounts[n]`). So user can also use appAccounts as well, instead of passing account address in appArgs.

Reply:
i think this file is written in TEALv2

Reply:
To write in *TEALv3*, we can just pass version = 3, on `compileTeal(..)`. But since we do not support it here i just suggested adding a note.